### ISSUE TITLE
Wrong values returned from getRadius(Vector2) method for HalfEllipse

### ISSUE THREAD

**Issue by TeamFinal** [dyn4j/dyn4j#45]
Hello William,
in the org.dyn4j.geometry.HalfEllipse class in the getRadius method you say:

> in the half ellipse case, if the point is on the side of the flat edge
> then we do the ellipse code, else we can just return the farthest of the
> two points that make up the flat side

But this is not true if the height is greater than half the width and the Point is below the [evolute](http://mathworld.wolfram.com/EllipseEvolute.html) of the ellipse. Then it is possible, that the point with the maximal distance is on the ellipse. But this point isn't on the opposite quadrant, as this is a half-ellipse, but on the neighbour quadrant. And to make things worse, there is as well a local minimum. Therefore the distance function isn't unimodal in this quadrant, which has to be for the Golden Section Search.

Thus I suggest to restrict the ellipse further:

![grafik](https://user-images.githubusercontent.com/46204212/50532243-e4f5a280-0b16-11e9-982e-9fe8cb6c3b14.png)

Greetings
Andreas

_Originally posted by @TeamFinal in https://github.com/dyn4j/dyn4j/issues/21#issuecomment-450253216_

**Comment by wnbittle** [comment]
You are correct.  Just so I understand, what did you want to restrict?  I think you are saying to restrict the height to less than or equal to the half width.  I think restricting by where the evolute + point are would be too confusing for the caller.

**Comment by TeamFinal** [comment]
I don't want to restrict the shape. All combinations of height and width of half ellipses should possible. The problem is if the given point is below the evolute, the max distance point may be on the ellipse and searching on the whole left part of the half ellipse with a golden section search isn't possible, because a golden section search needs an unimodal function. Therefore we have to restict the search to an interval in which the distance function is unimodal.
We know that the segment wih the local maximum length from the given point P to the ellipse (the orange segment) must be tangential to the evolute and the x-coordinate of the contact point B must smaller than that of the point P, while the segmnet with the local minimum length from the given point P to the ellipse  (the blue segment) must be on a line that is tangential to the evolute and that contact point must have a greater x-coordinate than P.
Thus the dotted tangente of the evolute splits the left quarter ellipse into two parts. The lower part with the local distance minimum and the upper part with the local distance maximum (the red part of the ellipse). In each part the distance function is unimodal. Thus a golden section search should be possible now.

This is the only idea I had to handle the issue, but I'm not very experianced in computational geometry. Thus there might be easily a much more elegant approach.

Greetings
Andreas

**Comment by wnbittle** [comment]
So does the slope of the evolute at the x coordinate of the given point alway split the ellipse into two unimodal regions?  Or only when the point lies below The Evolute?

**Comment by TeamFinal** [comment]
Let E(t)=(acos(t),bsin(t)) be a point on the ellipse and d: t -> ||E(t) - P|| the distance function for a given Point P.
Then d is a periodic function than may have one global and one local maximum between 0 and 2pi as well as one global and one local minimum. If the given point P is above (A) the evolute there should be only the global extrema, whith the distance function from a point below (B) should have all 4 extrema.

![image](https://user-images.githubusercontent.com/46204212/50815329-09266000-131d-11e9-9a61-cba52f414dad.png)

In the first case (point A) the code is already working.
In the second case it should be enough to get any one t_0 between the maximum and the minimum to split the distance function into tow unimodal functions. One defined from half pi to t_0 and the othe defined from from t_0 to pi.

For each of these E(t) the normals should go through the given point P and be tangential to the evolute.
Because of the monotony of these normals and the curvature of the evolute I'm pretty sure, that for all P that are below the evolute and in the 1st Quadrant. T_min has to be right below P and T_max left above.

![image](https://user-images.githubusercontent.com/46204212/50815186-89989100-131c-11e9-95d0-6272d7adf13a.png)

Pretty sure means I don't have any proofs nor do I have much experience. I didn't even know the term "evolute" before I started to find a solution for this problem.

**Comment by wnbittle** [comment]
No solution just yet, but wanted to post this graph so I don't lose it.

https://www.desmos.com/calculator/i5cva9llzp

**Comment by wnbittle** [comment]
I've checked in a update to the HalfEllipse class to solve this problem.  

From my testing I think you are correct - while the region split by the slope of the evolute at p.x is unimodal, the other region is not, but it doesn't matter because the maximal point is never in that region.

With that in mind, I've updated the code to compute whether the given point is below the evolute and in which case compute the bound to the unimodal region and perform a golden search there.  From my testing things look good so far.
